src/iceberg/util/snapshot_util_internal.h

```diff
@@ -0,0 +1,274 @@
+#include "iceberg/type_fwd.h"
+#include "iceberg/util/timepoint.h"
+
+namespace iceberg {
+
+/// \brief Utility functions for working with snapshots
+class ICEBERG_EXPORT SnapshotUtil {
+ public:
+  /// \brief Returns a vector of ancestors of the given snapshot.
+  ///
+  /// \param table The table
+  /// \param snapshot_id The snapshot ID to start from
+  /// \return A vector of ancestor snapshots
+  static Result<std::vector<std::shared_ptr<Snapshot>>> AncestorsOf(const Table& table,
+                                                                    int64_t snapshot_id);
+
+  /// \brief Returns whether ancestor_snapshot_id is an ancestor of snapshot_id.
+  ///
+  /// \param table The table to check
+  /// \param snapshot_id The snapshot ID to check
+  /// \param ancestor_snapshot_id The ancestor snapshot ID to check for
+  /// \return true if ancestor_snapshot_id is an ancestor of snapshot_id
+  static Result<bool> IsAncestorOf(const Table& table, int64_t snapshot_id,
+                                   int64_t ancestor_snapshot_id);
+
+  /// \brief Returns whether ancestor_snapshot_id is an ancestor of the table's current
+  /// state.
+  ///
+  /// \param table The table to check
+  /// \param ancestor_snapshot_id The ancestor snapshot ID to check for
+  /// \return true if ancestor_snapshot_id is an ancestor of the current snapshot
+  static Result<bool> IsAncestorOf(const Table& table, int64_t ancestor_snapshot_id);
+
+  /// \brief Returns whether some ancestor of snapshot_id has parentId matches
+  /// ancestor_parent_snapshot_id.
+  ///
+  /// \param table The table to check
+  /// \param snapshot_id The snapshot ID to check
+  /// \param ancestor_parent_snapshot_id The ancestor parent snapshot ID to check for
+  /// \return true if any ancestor has the given parent ID
+  static Result<bool> IsParentAncestorOf(const Table& table, int64_t snapshot_id,
+                                         int64_t ancestor_parent_snapshot_id);
+
+  /// \brief Returns a vector that traverses the table's snapshots from the current to the
+  /// last known ancestor.
+  ///
+  /// \param table The table
+  /// \return A vector from the table's current snapshot to its last known ancestor
+  static Result<std::vector<std::shared_ptr<Snapshot>>> CurrentAncestors(
+      const Table& table);
+
+  /// \brief Return the snapshot IDs for the ancestors of the current table state.
+  ///
+  /// Ancestor IDs are ordered by commit time, descending. The first ID is the current
+  /// snapshot, followed by its parent, and so on.
+  ///
+  /// \param table The table
+  /// \return A vector of snapshot IDs of the known ancestor snapshots, including the
+  /// current ID
+  static Result<std::vector<int64_t>> CurrentAncestorIds(const Table& table);
+
+  /// \brief Traverses the history of the table's current snapshot and finds the oldest
+  /// Snapshot.
+  ///
+  /// \param table The table
+  /// \return The oldest snapshot, or nullopt if there is no current snapshot
```

Comment:
It would be good to document that all returned `std::shared_ptr<Snapshot>` cannot be nullptr.